app/src/main/java/com/hedvig/app/feature/offer/ui/OfferActivity.kt

```diff
@@ -81,52 +81,39 @@ class OfferActivity : BaseActivity(R.layout.activity_offer) {
-                        storeBoolean(IS_VIEWING_OFFER, false)
-                        startActivity(
-                            Intent(
-                                this@OfferActivity,
-                                LoggedInActivity::class.java
-                            ).apply {
-                                putExtra(LoggedInActivity.EXTRA_IS_FROM_ONBOARDING, true)
-                                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
-                                addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
-                            }
-                        )
-                    } else {
-                        (offerScroll.adapter as? OfferAdapter)?.submitList(
-                            listOfNotNull(
-                                OfferModel.Header(data),
-                                OfferModel.Info,
-                                OfferModel.Facts(data),
-                                OfferModel.Perils(data),
-                                OfferModel.Terms(data),
-                                data.lastQuoteOfMember.asCompleteQuote?.currentInsurer?.let { currentInsurer ->
-                                    if (currentInsurer.switchable == true) {
-                                        OfferModel.Switcher(currentInsurer.displayName)
-                                    } else {
-                                        null
-                                    }
-                                },
-                                OfferModel.Footer
-                            )
-                        )
-                    }
+            val adapter = OfferAdapter(
+                fragmentManager = supportFragmentManager,
+                tracker = tracker,
+                marketManager = marketManager,
+                removeDiscount = model::removeDiscount
+            )
+            offerScroll.adapter = adapter
+
+            model.viewState.observe(this@OfferActivity) { viewState ->
+                when (viewState) {
+                    OfferViewModel.ViewState.HasContracts -> startLoggedInActivity()
+                    is OfferViewModel.ViewState.OfferItems -> adapter.submitList(viewState.items)
+                    is OfferViewModel.ViewState.Error.GeneralError -> showErrorDialog(viewState.message ?: getString(R.string.home_tab_error_body))
+                    is OfferViewModel.ViewState.Error -> showErrorDialog(getString(R.string.home_tab_error_body))
                 }
             }
         }
     }
 
+    private fun startLoggedInActivity() {
+        storeBoolean(IS_VIEWING_OFFER, false)
+        startActivity(
+            Intent(
+                this@OfferActivity,
+                LoggedInActivity::class.java
+            ).apply {
+                putExtra(LoggedInActivity.EXTRA_IS_FROM_ONBOARDING, true)
+                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+                addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
+            }
```

Comment:
Could we move this `Intent`-creation to use `LoggedInActivity::newInstance`? I think that would limit the amount of `Intent`-parameters to an amount that is easier to reason about 🙂